pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -665,22 +948,107 @@ func (b *binaryOperationSeriesBuffer) close() {
 
+// FIXME(jhesketh): Investigate avoiding copying histograms for binary ops.
 var arithmeticOperationFuncs = map[parser.ItemType]binaryOperationFunc{
-	parser.ADD: func(left, right float64) float64 {
-		return left + right
+	parser.ADD: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Add(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs + rhs, nil, true, nil
+	},
+	parser.SUB: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Sub(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs - rhs, nil, true, nil
+	},
+	parser.MUL: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs == nil {
+			return 0, hlhs.Copy().Mul(rhs), true, nil
+		}
+		if hlhs == nil && hrhs != nil {
+			return 0, hrhs.Copy().Mul(lhs), true, nil
+		}
+		return lhs * rhs, nil, true, nil
 	},
-	parser.SUB: func(left, right float64) float64 {
-		return left - right
+	parser.DIV: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs == nil {
+			return 0, hlhs.Copy().Div(rhs), true, nil
+		}
+		return lhs / rhs, nil, true, nil
 	},
-	parser.MUL: func(left, right float64) float64 {
-		return left * right
+	parser.POW: func(lhs, rhs float64, _, _ *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Pow(lhs, rhs), nil, true, nil
 	},
-	parser.DIV: func(left, right float64) float64 {
-		return left / right
+	parser.MOD: func(lhs, rhs float64, _, _ *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Mod(lhs, rhs), nil, true, nil
 	},
-	parser.MOD:   math.Mod,
-	parser.POW:   math.Pow,
-	parser.ATAN2: math.Atan2,
+	parser.ATAN2: func(lhs, rhs float64, _, _ *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Atan2(lhs, rhs), nil, true, nil
+	},
+}
+
+// seriesForOneGroupSide couples data and sourceSeriesIndices for use by mergeOneSide.
+// Specifically it is used for histogramSideSorter and floatSideSorter which providing a sorting interface
+// (sort.Interface) which will sort both data and sourceSeriesIndices at the same time to retain a mapping
+// of their indexes on either the present float timestamps or histogram timestamps.
+type seriesForOneGroupSide struct {
+	data                []types.InstantVectorSeriesData
+	sourceSeriesIndices []int
+}
+
+type floatSideSorter struct {
```

Comment:
[nit] Might be good to add a comment explaining what `floatSideSorter` does (eg. "sorts side by the timestamp of the first float point in each series")